### PR TITLE
Fuzzy dates

### PR DESCRIPTION
**Pull request by ayco-at-naturalis**

### CODE REVIEW

File: colplus-admin/src/main/java/org/col/admin/task/importer/InterpreterBase.java [hunk 1]
```diff
@@ -74,14 +82,27 @@ protected void addAndTransliterate(NeoTaxon t, VernacularName vn) {
   }
 
   protected LocalDate date(NeoTaxon t, Issue invalidIssue, Term term) {
-    return parse(DateParser.PARSER, t.verbatim.getTerm(term)).orNull(invalidIssue, t.issues);
+    Optional<FuzzyDate> date;
```
Comment (mdoering) [inline colplus-admin/src/main/java/org/col/admin/task/importer/InterpreterBase.java]: this try catch and add issue is done with the help of the SafeParser in the "parse" method already, better reuse that and just do the if fuzzy data add partial date issue

File: colplus-admin/src/main/java/org/col/admin/task/importer/InterpreterBase.java [hunk 2]
```diff
@@ -151,12 +172,11 @@ public Name interpretName(String id, String vrank, String sciname, String author
...
-          if (!n.authorshipComplete().equalsIgnoreCase(pnAuthorship.authorshipComplete())){
+          if (!n.authorshipComplete().equalsIgnoreCase(pnAuthorship.authorshipComplete())) {
             n.addIssue(Issue.INCONSISTENT_AUTHORSHIP);
-            LOG.info("Different authorship found in dwc:scientificName=[{}] and dwc:scientificNameAuthorship=[{}]",
-                n.authorshipComplete(),
-                pnAuthorship.authorshipComplete()
-            );
+            LOG.info(
+                "Different authorship found in dwc:scientificName=[{}] and dwc:scientificNameAuthorship=[{}]",
+                n.authorshipComplete(), pnAuthorship.authorshipComplete());
```
Comment (mdoering) [inline colplus-admin/src/main/java/org/col/admin/task/importer/InterpreterBase.java]: do you prefer this style?

File: colplus-admin/src/main/java/org/col/admin/task/importer/acef/AcefDateParser.java [hunk 3]
```diff
@@ -1,86 +1,13 @@
...
-import java.util.List;
-import java.util.Optional;
-
-/*
- * Currently allows only exactly for those patterns actually encountered in the ACEF sources. We
- * could allow for more date patterns (e.g. month not case sensitive), or even allow the patterns to
- * be settable.
- */
-public class AcefDateParser implements Parser<LocalDate> {
+public class AcefDateParser extends FuzzyDateParser {
```
Comment (mdoering) [inline colplus-admin/src/main/java/org/col/admin/task/importer/acef/AcefDateParser.java]: Is it worth a class on its own?
I would have rather thought we have a single DateParser class with some specific factory methods or singletons that load the appropriate configs. Sth like:

class  DateParser {
  public static final ACEF = new DateParser("acef.fuzzy-date.properties")

File: colplus-admin/src/test/java/org/col/admin/task/importer/acef/AcefDateParserTest.java [hunk 4]
```diff
@@ -1,66 +1,36 @@
...
   }
 
-  @Test
-  public void testUnparsable() throws Exception {
-    for (String x : new String[]{"321", "jan", "31-01"}) {
-      assertUnparsable(x);
-    }
+  public void test3() throws UnparsableException {
+    Optional<FuzzyDate> fd = AcefDateParser.PARSER.parse("Nov-2010");
+    assertTrue("01", fd.isPresent());
```
Comment (mdoering) [inline colplus-admin/src/test/java/org/col/admin/task/importer/acef/AcefDateParserTest.java]: what is "01" used for? Is it just identifying in logs where a test has failed? I don't find this very useful, rather confusing. Can we either use real messages or none?

File: colplus-admin/src/main/java/org/col/admin/task/importer/InterpreterBase.java (hunk 1 above)
Comment (ayco-at-naturalis) [reply colplus-admin/src/main/java/org/col/admin/task/importer/InterpreterBase.java]: Yes, my problem was: the SafeParser allows for only 1 issue to be reported (INVALID_DATE) __if__ we also like to report (warning) that the date partial a.k.a. fuzzy, I couldn't do it via the SafeParser. Let me know if you think it is important that we found a fuzzy date. If not, everything can be folded back into the SafeParser again

Comment (ayco-at-naturalis) [reply colplus-admin/src/main/java/org/col/admin/task/importer/InterpreterBase.java]: (I have created a new Issue constant: PARTIAL_DATE)

**Comment by ayco-at-naturalis** [comment]
I think I have found a way to decouple generic date parsing logic from data source specific logic. That allows us to add more data sources, or even create parsers for specific fields. I will push that. Let me know if you think that's better

**Comment by mdoering** [comment]
The NPE problem here seems still valid to me: https://github.com/Sp2000/colplus-backend/commit/cc1846c3f2000fcdc1ce3643bf91dbf7e876b5d7#r28266928

Otherwise good to be merged back!

**Pull request merged**